pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -478,535 +262,50 @@ func (fs *Decomposedfs) PurgeExpiredUploads(purgedChan chan<- tusd.FileInfo) err
-func (upload *fileUpload) Terminate(ctx context.Context) error {
-	if err := os.Remove(upload.infoPath); err != nil {
-		if !errors.Is(err, iofs.ErrNotExist) {
-			return err
-		}
-	}
-	if err := os.Remove(upload.binPath); err != nil {
-		if !errors.Is(err, iofs.ErrNotExist) {
-			return err
-		}
-	}
-	return nil
+// To implement the termination extension as specified in https://tus.io/protocols/resumable-upload.html#termination
+// the storage needs to implement AsTerminatableUpload
+func (fs *Decomposedfs) AsTerminatableUpload(up tusd.Upload) tusd.TerminatableUpload {
+	return up.(*upload.Upload)
 }
 
-// To implement the creation-defer-length extension as specified in https://tus.io/protocols/resumable-upload.html#creation
-// - the storage needs to implement AsLengthDeclarableUpload
-// - the upload needs to implement DeclareLength
-
 // AsLengthDeclarableUpload returns a LengthDeclarableUpload
-func (fs *Decomposedfs) AsLengthDeclarableUpload(upload tusd.Upload) tusd.LengthDeclarableUpload {
-	return upload.(*fileUpload)
-}
-
-// DeclareLength updates the upload length information
-func (upload *fileUpload) DeclareLength(ctx context.Context, length int64) error {
-	upload.info.Size = length
-	upload.info.SizeIsDeferred = false
-	return upload.writeInfo()
+// To implement the creation-defer-length extension as specified in https://tus.io/protocols/resumable-upload.html#creation
+// the storage needs to implement AsLengthDeclarableUpload
+func (fs *Decomposedfs) AsLengthDeclarableUpload(up tusd.Upload) tusd.LengthDeclarableUpload {
+	return up.(*upload.Upload)
 }
 
-// To implement the concatenation extension as specified in https://tus.io/protocols/resumable-upload.html#concatenation
-// - the storage needs to implement AsConcatableUpload
-// - the upload needs to implement ConcatUploads
-
 // AsConcatableUpload returns a ConcatableUpload
-func (fs *Decomposedfs) AsConcatableUpload(upload tusd.Upload) tusd.ConcatableUpload {
-	return upload.(*fileUpload)
+// To implement the concatenation extension as specified in https://tus.io/protocols/resumable-upload.html#concatenation
+// the storage needs to implement AsConcatableUpload
+func (fs *Decomposedfs) AsConcatableUpload(up tusd.Upload) tusd.ConcatableUpload {
+	return up.(*upload.Upload)
 }
 
-// ConcatUploads concatenates multiple uploads
-func (upload *fileUpload) ConcatUploads(ctx context.Context, uploads []tusd.Upload) (err error) {
-	file, err := os.OpenFile(upload.binPath, os.O_WRONLY|os.O_APPEND, defaultFilePerm)
+func (fs *Decomposedfs) uploadInfos(ctx context.Context) ([]tusd.FileInfo, error) {
+	infos := []tusd.FileInfo{}
+	infoFiles, err := filepath.Glob(filepath.Join(fs.o.Root, "uploads", "*.info"))
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer file.Close()
-
-	for _, partialUpload := range uploads {
-		fileUpload := partialUpload.(*fileUpload)
 
-		src, err := os.Open(fileUpload.binPath)
+	idRegexp := regexp.MustCompile(".*/([^/]+).info")
```

Comment:
This can/should be moved to a global variable so that we don't recompile the same pattern over and over again.

Reply:
Excellently spotted :+1: Moved to global var